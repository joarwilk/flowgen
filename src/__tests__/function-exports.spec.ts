import { compiler, beautify } from "..";
import "../test-matchers";

it("should handle exported es module functions", () => {
  const ts = `
class Action {}
class RouterState {}
class Store<T> {}
class SyncHistoryWithStoreOptions {}
class HistoryUnsubscribe {}

export function routerReducer(state?: RouterState, action?: Action): RouterState;
export function syncHistoryWithStore(history: History, store: Store<any>, options?: SyncHistoryWithStoreOptions): History & HistoryUnsubscribe;
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle toString function overload", () => {
  const ts = `export function toString(): void;
export function toString(e: number): void;
export function toString(b: string): void;
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle default exported es module functions", () => {
  const ts = `
class Action {}
class RouterState {}

export default function routerReducer(state?: RouterState, action?: Action): RouterState;`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should handle function overload es module functions", () => {
  const ts = `
class Action {}
class RouterState {}

export function routerReducer(state?: RouterState, action?: Action): RouterState;
export function routerReducer(state?: RouterState): RouterState;
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should keep this annotation in functions", () => {
  const ts =
    "function addClickListener(onclick: (this: void, e: Event) => void): void;";
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should keep default parameters in functions", () => {
  const ts =
    "function addClickListener<T = Error>(onclick: (e: Event) => void): T;";
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should transform generic `<T extends E>` to `<T: E>` on functions", () => {
  const ts =
    "function addClickListener<T extends Error>(onclick: (e: Event) => void): T;";
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should not break with Promise return types - issue 156", () => {
  const ts = "export declare const fn: () => Promise<void>;";
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});
