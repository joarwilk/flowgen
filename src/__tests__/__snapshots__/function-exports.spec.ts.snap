// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle default exported es module functions 1`] = `
"declare class Action {}
declare class RouterState {}
declare export default function routerReducer(
  state?: RouterState,
  action?: Action
): RouterState;
"
`;

exports[`should handle exported es module functions 1`] = `
"declare class Action {}
declare class RouterState {}
declare class Store<T> {}
declare class SyncHistoryWithStoreOptions {}
declare class HistoryUnsubscribe {}
declare export function routerReducer(
  state?: RouterState,
  action?: Action
): RouterState;
declare export function syncHistoryWithStore(
  history: History,
  store: Store<any>,
  options?: SyncHistoryWithStoreOptions
): History & HistoryUnsubscribe;
"
`;

exports[`should handle function overload es module functions 1`] = `
"declare class Action {}
declare class RouterState {}
declare export function routerReducer(
  state?: RouterState,
  action?: Action
): RouterState;
declare export function routerReducer(state?: RouterState): RouterState;
"
`;

exports[`should handle toString function overload 1`] = `
"declare export function toString(): void;
declare export function toString(e: number): void;
declare export function toString(b: string): void;
"
`;

exports[`should keep default parameters in functions 1`] = `
"declare function addClickListener<T = Error>(onclick: (e: Event) => void): T;
"
`;

exports[`should keep this annotation in functions 1`] = `
"declare function addClickListener(
  onclick: (this: void, e: Event) => void
): void;
"
`;

exports[`should not break with Promise return types - issue 156 1`] = `
"declare export var fn: () => Promise<void>;
"
`;

exports[`should transform generic \`<T extends E>\` to \`<T: E>\` on functions 1`] = `
"declare function addClickListener<T: Error>(onclick: (e: Event) => void): T;
"
`;
